File: process-sample/src/main/scala/pl/touk/esp/process/sample/TestProcessConfigCreator.scala
package pl.touk.esp.process.sample

import com.typesafe.config.Config
import org.apache.flink.api.common.ExecutionConfig
import org.apache.flink.streaming.api.scala._

import pl.touk.esp.engine.api.process.{ProcessConfigCreator, SinkFactory, SourceFactory}
import pl.touk.esp.engine.api.{SkipExceptionHandler, MetaData, Service}
import pl.touk.esp.engine.process.util.CollectionSource
import pl.touk.esp.engine.util.sink.ServiceSink

import scala.concurrent.duration._
import scala.concurrent.{ExecutionContext, Future}

class TestProcessConfigCreator extends ProcessConfigCreator {

  override def sinkFactories(config: Config) = {
    val sendSmsSink = new ServiceSink(EmptyService, invocationTimeout = 2 minutes)
    val monitorSink = new ServiceSink(EmptyService, invocationTimeout = 2 minutes)
    Map[String, SinkFactory](
      "sendSms" -> SinkFactory.noParam(sendSmsSink),
      "monitor" -> SinkFactory.noParam(monitorSink)
    )
  }

  override def listeners(config: Config) = List()

  override def sourceFactories(config: Config) = {
    Map[String, SourceFactory[_]](
      "kafka-transaction" -> new SourceFactory[String] {
        override def create(processMetaData: MetaData, parameters: Map[String, String]) = {
          CollectionSource(new ExecutionConfig(), List("blee"), None)
        }
      }
    )
  }

  override def services(config: Config) = {
    Map[String, Service](
      "accountService" -> EmptyService,
      "componentService" -> EmptyService,
      "transactionService" -> EmptyService,
      "serviceModelService" -> EmptyService
    )
  }

  override def foldingFunctions(config: Config) = Map()

  override def exceptionHandler(config: Config) = SkipExceptionHandler
}

case object EmptyService extends Service {
  override def invoke(params: Map[String, Any])(implicit ec: ExecutionContext) = {
    Future(println(params))
  }
}

File: engine/flink/management/dev-model-java/src/main/java/pl/touk/nussknacker/engine/management/javasample/TestProcessConfigCreator.java
package pl.touk.nussknacker.engine.management.javasample;

import pl.touk.nussknacker.engine.api.CustomStreamTransformer;
import pl.touk.nussknacker.engine.api.ProcessListener;
import pl.touk.nussknacker.engine.api.Service;
import pl.touk.nussknacker.engine.api.process.ProcessObjectDependencies;
import pl.touk.nussknacker.engine.api.process.SinkFactory;
import pl.touk.nussknacker.engine.api.process.SourceFactory;
import pl.touk.nussknacker.engine.api.process.WithCategories;
import pl.touk.nussknacker.engine.javaapi.process.ExpressionConfig;
import pl.touk.nussknacker.engine.javaapi.process.ProcessConfigCreator;

import java.util.Collection;
import java.util.Collections;
import java.util.Map;

public class TestProcessConfigCreator implements ProcessConfigCreator {

    private Objects objects = new Objects();

    @Override
    public Map<String, WithCategories<Service>> services(ProcessObjectDependencies processObjectDependencies) {
        return Collections.emptyMap();
    }

    @Override
    public Map<String, WithCategories<SourceFactory>> sourceFactories(ProcessObjectDependencies processObjectDependencies) {
        return Collections.singletonMap("source", objects.source());
    }

    @Override
    public Map<String, WithCategories<SinkFactory>> sinkFactories(ProcessObjectDependencies processObjectDependencies) {
        return Collections.singletonMap("sink", objects.sink());
    }

    @Override
    public Map<String, WithCategories<CustomStreamTransformer>> customStreamTransformers(ProcessObjectDependencies processObjectDependencies) {
        return Collections.emptyMap();
    }

    @Override
    public Collection<ProcessListener> listeners(ProcessObjectDependencies processObjectDependencies) {
        return Collections.emptyList();
    }

    @Override
    public ExpressionConfig expressionConfig(ProcessObjectDependencies processObjectDependencies) {
        return new ExpressionConfig(Collections.emptyMap(), Collections.emptyList());
    }

    @Override
    public Map<String, String> modelInfo() {
        return Collections.emptyMap();
    }
}
